kvm/src/lib.rs

```diff
@@ -1369,20 +1369,22 @@ mod tests {
 
     #[cfg(target_arch = "x86_64")]
     #[test]
-    fn run_code_test() {
+    fn test_run_code() {
         use std::io::Write;
 
         let kvm = Kvm::new().unwrap();
         let vm = kvm.create_vm().unwrap();
         // This example based on https://lwn.net/Articles/658511/
+        #[rustfmt::skip]
         let code = [
             0xba, 0xf8, 0x03, /* mov $0x3f8, %dx */
             0x00, 0xd8, /* add %bl, %al */
             0x04, b'0', /* add $'0', %al */
             0xee, /* out %al, %dx */
             0xec, /* in %dx, %al */
-            0xc6, 0x06, 0x00, 0x20, 0x00, /* movl $0, (0x2000) */
-            0x8a, 0x16, 0x00, 0x20, /* movl (0x2000), %dl */
+            0xc6, 0x06, 0x00, 0x80, 0x00, /* movl $0, (0x8000); This generates a MMIO Write.*/
+            0x8a, 0x16, 0x00, 0x80, /* movl (0x8000), %dl; This generates a MMIO Read.*/
+            0xc6, 0x06, 0x00, 0x20, 0x00, /* movl $0, (0x2000); Dirty one page in guest mem. */
```

Comment:
If this code dirties one page in the guest, should the  `assert_eq!(dirty_pages, 2);` from this test fail?

Reply:
The second page is dirtied by the code itself when it gets loaded in memory.

Reply:
My confusion is triggered by the fact that this line is added in this PR (from what I see it did not exist before, am I wrong?) while the assert was there before that. Question is why did the assert hold before and why does it still hold now with no change?

Reply:
Also, this is ASM magic and the provided link
> // This example based on https://lwn.net/Articles/658511/

didn't clarify anything for me. Could you add more detail on what this ASM is supposed to do?

Reply:
The assert related to the number of kvm_dirty_log was correct before because initially we were dirtying 2 pages unintentionally:
- one page was dirtied by loading the code in guest memory (this still happens)
- the second page was wrongfully dirtied by this line: `0xc6, 0x06, 0x00, 0x20, 0x00,`. This is the bug that I am fixing in this PR. This instruction is a write at the address `0x2000` which was added in the first version of this test to check that MMIOWrite is correct. Only that because 0x2000 is an address inside the guest memory address space, that instruction does not trigger MMIO, but a classic write in guest memory, hence there are 2 dirty pages. The change that introduce this bug was the one where we increase the `mem_size` from 0x1000 to 0x40000. That happened a while back and in the meantime we changed that `mem_size` without knowing that it influences the MMIO.

Now, to have an additional write (besides the code write) in the guest memory that increases the dirty pages count to 2, I added a new write instruction:
`0xc6, 0x06, 0x00, 0x20, 0x00,`

To fix the test in terms of triggering MMIO Reads and MMIO Writes, I changes the address from 0x2000 (address inside guest memory region) to 0x8000 (address outside of the guest memory region). If the address is outside of the guest memory region, it will trigger the MMIO thus checking the exists MMIORead & MMIOWrite.

The commit that adds the MMIO checks is 97a8816486fe315812d605d4f3fb11ddb3231c8c.
Here is the initial working code (before adding kvm_dirty_log), maybe this will clear up some things:
```
 // This example based on https://lwn.net/Articles/658511/
        let code = [
            0xba, 0xf8, 0x03 /* mov $0x3f8, %dx */, 0x00, 0xd8 /* add %bl, %al */, 0x04,
            '0' as u8 /* add $'0', %al */, 0xee /* out %al, %dx */,
            0xec /* in %dx, %al */, 0xc6, 0x06, 0x00, 0x20, 0x00 /* movl $0, (0x2000) */,
            0x8a, 0x16, 0x00, 0x20 /* movl (0x2000), %dl */, 0xf4 /* hlt */,
        ];
....
loop {
            match vcpu_fd.run().expect("run failed") {
                VcpuExit::IoIn(addr, data) => {
                    assert_eq!(addr, 0x3f8);
                    assert_eq!(data.len(), 1);
                }
                VcpuExit::IoOut(addr, data) => {
                    assert_eq!(addr, 0x3f8);
                    assert_eq!(data.len(), 1);
                    assert_eq!(data[0], '5' as u8);
                }
                VcpuExit::MmioRead(addr, data) => {
                    assert_eq!(addr, 0x2000);
                    assert_eq!(data.len(), 1);
                }
                VcpuExit::MmioWrite(addr, data) => {
                    assert_eq!(addr, 0x2000);
                    assert_eq!(data.len(), 1);
                    assert_eq!(data[0], 0);
                }
                VcpuExit::Hlt => {
                    break;
                }
                r => panic!("unexpected exit reason: {:?}", r),
            }
        }
```
I also don't believe that the purpose of the test is to explain in detail the ASM. The basics are explained in the provided link.

Reply:
**Now, to have an additional write (besides the code write) in the guest memory that increases the dirty pages count to 2, I added a new write instruction:
0xc6, 0x06, 0x00, 0x20, 0x00** -> this explains it all. You kept around the `0x20`  write to trigger a second dirty page, that is what I did not get initially.
I do not see the too much value in triggering a random write just to have 2 dirty pages instead of 1 but I am not very keen on it.